File: frontend/src/lib/driving.ts
/**
 * What is driving each parameter of the rig, as the evaluator wants it.
 *
 * The console keeps *what is driving* a parameter — the fade or shape anchored in
 * console time, the programmer over it, the home value beneath — and nobody keeps the
 * answer. This turns the rows a browser already has into the four layers the evaluator
 * takes, and it is all this file does: there is no arithmetic here, deliberately.
 * Turning those layers into a number happens in one place, in Rust, compiled for both
 * the station and the page.
 */

import type {
	Fixture,
	FixtureType,
	ParameterValue,
	ProgrammerValue,
	RunningEffect,
	RunningFade
} from './generated/index.js';
import { parameterKey } from './patch.js';

/** What the evaluator is given for one parameter. */
export type DrivenBy = {
	/** A plain value the programmer is holding. A held *shape* is not here: the
	 *  station resolves it against its speed master and publishes it as `effect`. */
	programmer?: ParameterValue;
	effect?: RunningEffect;
	fade?: RunningFade;
	home?: ParameterValue;
};

/** How a parameter is named everywhere the evaluator is concerned. */
export const drivingKey = (fixtureId: string, key: string) => `${fixtureId}/${key}`;

/**
 * Where a parameter rests when nothing is driving it: this fixture's own override, or
 * what its type declares.
 *
 * The same resolution `home_value_by_key` makes on the station, and it has to stay the
 * same — a browser that invented its own answer would draw an untouched house light
 * dark while the station had it up.
 */
export function homeValue(
	fixture: Fixture,
	type: FixtureType | undefined,
	key: string
): ParameterValue | undefined {
	const own = fixture.home_values[key];
	if (own !== undefined) return own;
	return type?.parameters.find((p) => parameterKey(p.kind) === key)?.default_value;
}

/**
 * Everything acting on one fixture, keyed by parameter key.
 *
 * Only the parameters something can say anything about: a key with no fade, no shape,
 * no hold and no home is a key nothing has ever driven and nothing can place, and
 * listing it would only make the evaluator answer "nothing" more often.
 */
export function drivenBy(
	fixture: Fixture,
	type: FixtureType | undefined,
	held: Map<string, ProgrammerValue>
): Map<string, DrivenBy> {
	const rows = new Map<string, DrivenBy>();
	const keys = new Set<string>([
		...Object.keys(fixture.live_effects),
		...Object.keys(fixture.live_fades),
		...Object.keys(fixture.home_values),
		...(type?.parameters ?? []).map((p) => parameterKey(p.kind))
	]);

	for (const key of keys) {
		const entry = held.get(drivingKey(fixture.id, key));
		const row: DrivenBy = {};
		// An entry carrying a shape asserts the shape, never a value: grabbing a fader
		// and putting a sine on it are the same act of taking hold of one parameter.
		if (entry && !entry.effect) row.programmer = entry.value;
		const effect = fixture.live_effects[key];
		if (effect) row.effect = effect;
		const fade = fixture.live_fades[key];
		if (fade) row.fade = fade;
		const home = homeValue(fixture, type, key);
		if (home !== undefined) row.home = home;
		if (Object.keys(row).length > 0) rows.set(key, row);
	}
	return rows;
}

/** The whole rig, flattened into the map the evaluator holds. */
export function drivingTheRig(
	fixtures: Fixture[],
	types: FixtureType[],
	entries: ProgrammerValue[]
): Record<string, DrivenBy> {
	const byId = new Map(types.map((t) => [t.id, t]));
	const held = new Map<string, ProgrammerValue>();
	for (const entry of entries) {
		held.set(drivingKey(entry.fixture_id, parameterKey(entry.parameter_kind)), entry);
	}

	const out: Record<string, DrivenBy> = {};
	for (const fixture of fixtures) {
		for (const [key, row] of drivenBy(fixture, byId.get(fixture.fixture_type_id), held)) {
			out[drivingKey(fixture.id, key)] = row;
		}
	}
	return out;
}

File: crates/pult-render/src/driving.rs
//! What is driving one parameter, and therefore what it is putting out.
//!
//! The console keeps *what is driving* a parameter — a fade anchored in time, an
//! effect anchored in time, the programmer over the top, the home value underneath —
//! and nobody keeps the answer. This is the function that turns the first into the
//! second, and it is the whole of the priority rule: the programmer wins, then an
//! effect, then a fade, then where the parameter rests.
//!
//! The order matters less than it looks, because the station only ever publishes the
//! winner of the two middle layers: a fade under an effect is not listed at all. What
//! the order is really for is the programmer, which is published separately because
//! it is SYNCED show data rather than something a station worked out.
//!
//! A fade that has arrived is deliberately still a fade. It is the only record of
//! where the parameter got to — nothing stores the number any more — and evaluating a
//! finished fade is exactly the constant it landed on.

use crate::effect::{fade_value_at, effect_value_at, RunningEffect, RunningFade};
use crate::value::ParameterValue;

/// The layers acting on one parameter, highest priority first.
///
/// Every field is optional and all four may be absent, which is a parameter nothing
/// has ever driven and whose type declares no default — the one case that has no
/// value at all.
#[derive(Debug, Default, Clone, Copy)]
pub struct Driving<'a> {
    /// A plain value the programmer is holding. A programmer *effect* is not here:
    /// the station resolves it against its speed master and publishes it as the
    /// running effect below, which is what keeps rate-following out of this crate.
    pub programmer: Option<&'a ParameterValue>,
    pub effect: Option<&'a RunningEffect>,
    pub fade: Option<&'a RunningFade>,
    pub home: Option<&'a ParameterValue>,
}

impl<'a> Driving<'a> {
    /// True when some layer is asserting something, home or not.
    pub fn is_empty(&self) -> bool {
        self.programmer.is_none()
            && self.effect.is_none()
            && self.fade.is_none()
            && self.home.is_none()
    }

    /// True when playback or the programmer is asserting something — that is, when
    /// this parameter is being driven rather than merely resting.
    pub fn is_driven(&self) -> bool {
        self.programmer.is_some() || self.effect.is_some() || self.fade.is_some()
    }
}

/// What this parameter is putting out at `now_ms`.
///
/// `None` only where nothing at all applies, which a caller reads as "this fixture
/// has no such parameter" rather than as a zero of some shape.
pub fn value_at(driving: &Driving<'_>, now_ms: u64) -> Option<ParameterValue> {
    if let Some(held) = driving.programmer {
        return Some(held.clone());
    }
    if let Some(effect) = driving.effect {
        return Some(effect_value_at(effect, now_ms));
    }
    if let Some(fade) = driving.fade {
        return Some(fade_value_at(fade, now_ms));
    }
    driving.home.cloned()
}

/// The console millisecond after which this parameter stops changing on its own, or
/// `None` if it never does.
///
/// What a consumer with a frame rate asks in order to stop paying for frames nobody
/// can see a difference in. An effect runs for ever, so it answers `None`; a fade
/// answers when it lands; a held value and a home value answer "already".
pub fn settles_at(driving: &Driving<'_>) -> Option<u64> {
    if driving.programmer.is_some() {
        return Some(0);
    }
    if driving.effect.is_some() {
        return None;
    }
    match driving.fade {
        Some(fade) => Some(fade.t0.saturating_add(fade.duration_ms as u64)),
        None => Some(0),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::effect::{Curve, Direction, Easing, EffectSource, Shape};
    use uuid::Uuid;

    fn a_fade(from: f32, to: f32, t0: u64, duration_ms: u32) -> RunningFade {
        RunningFade {
            from: ParameterValue::Float(from),
            to: ParameterValue::Float(to),
            t0,
            duration_ms,
            easing: Easing::Linear,
            cue_id: Uuid::nil(),
        }
    }

    fn an_effect(t0: u64) -> RunningEffect {
        RunningEffect {
            effect_id: Uuid::nil(),
            curve: Curve::Shape(Shape::SawUp),
            rate_hz: 1.0,
            low: ParameterValue::Float(0.0),
            high: ParameterValue::Float(1.0),
            width: 0.5,
            direction: Direction::Forward,
            phase: 0.0,
            t0,
            source: EffectSource::Programmer,
        }
    }

    #[test]
    fn nothing_driving_a_parameter_gives_its_home_value() {
        let home = ParameterValue::Float(0.25);
        let driving = Driving { home: Some(&home), ..Default::default() };
        assert_eq!(value_at(&driving, 1_000), Some(ParameterValue::Float(0.25)));
        assert!(!driving.is_driven());
    }

    #[test]
    fn a_parameter_nothing_can_say_anything_about_has_no_value() {
        assert_eq!(value_at(&Driving::default(), 1_000), None);
        assert!(Driving::default().is_empty());
    }

    #[test]
    fn a_fade_beats_the_home_value_and_moves_with_the_moment() {
        let home = ParameterValue::Float(0.0);
        let fade = a_fade(0.0, 1.0, 1_000, 1_000);
        let driving = Driving { fade: Some(&fade), home: Some(&home), ..Default::default() };
        assert_eq!(value_at(&driving, 1_500), Some(ParameterValue::Float(0.5)));
    }

    #[test]
    fn a_fade_that_has_arrived_is_still_where_the_parameter_is() {
        let fade = a_fade(0.0, 0.8, 1_000, 500);
        let driving = Driving { fade: Some(&fade), ..Default::default() };
        assert_eq!(value_at(&driving, 9_999_999), Some(ParameterValue::Float(0.8)));
    }

    #[test]
    fn an_effect_beats_a_fade_and_the_programmer_beats_both() {
        let fade = a_fade(0.0, 1.0, 0, 1_000);
        let effect = an_effect(0);
        let held = ParameterValue::Float(0.125);

        let over_fade = Driving { effect: Some(&effect), fade: Some(&fade), ..Default::default() };
        assert_eq!(value_at(&over_fade, 250), Some(ParameterValue::Float(0.25)));

        let over_everything = Driving { programmer: Some(&held), ..over_fade };
        assert_eq!(value_at(&over_everything, 250), Some(ParameterValue::Float(0.125)));
    }

    #[test]
    fn what_settles_says_when_there_is_nothing_more_to_draw() {
        let fade = a_fade(0.0, 1.0, 1_000, 400);
        let effect = an_effect(0);
        assert_eq!(settles_at(&Driving { fade: Some(&fade), ..Default::default() }), Some(1_400));
        assert_eq!(settles_at(&Driving { effect: Some(&effect), ..Default::default() }), None);
        assert_eq!(settles_at(&Driving::default()), Some(0));
    }
}
